aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -765,7 +872,7 @@ static void apply_geqrf(Tensor& self, Tensor& tau, int64_t m, int64_t n,
   int lwork = -1;
   scalar_t wkopt;
   lapackGeqrf<scalar_t>(m, n, self_data, m, tau_data, &wkopt, lwork, &info);
-  lwork = static_cast<int>(wkopt);
+  lwork = static_cast<int>(zabs<scalar_t, value_t>(wkopt));
```

Comment:
I thought about it a while and couldn't figure it out. Why is the zabs call here necessary?

Reply:
wkopt cannot be cast directly to the lwork data type without converting to a real number. wkopt is a positive real number that is stored in a std::complex variable thus `zabs()` and `real_impl()` should both do the job, but `real_impl()` is more performant.

I have updated the template utility functions in `zmath.h` to support:
 - complex return type: when performing tensor operations that maintain a fixed dtype.
 - real return type: when performing low-level C++ operations.
All of these functions are no-ops for other data types.